File: src/main/scala/io/boson/bson/bsonImpl/BosonFuse.scala
package io.boson.bson.bsonImpl

import java.nio.ByteBuffer
import java.util.concurrent.CompletableFuture

import io.boson.bson
import io.boson.bson.Boson

class BosonFuse(first: bson.Boson, second: bson.Boson) extends bson.Boson {

  override def go(bsonByteEncoding: Array[Byte]): CompletableFuture[Array[Byte]] = {
    val future: CompletableFuture[Array[Byte]] =
      CompletableFuture.supplyAsync(() => {
            val firstFuture: CompletableFuture[Array[Byte]] = first.go(bsonByteEncoding)
            second.go(firstFuture.join()).join()
        })
        future
  }

  override def go(bsonByteBufferEncoding: ByteBuffer): CompletableFuture[ByteBuffer] = {
    val future: CompletableFuture[ByteBuffer] =
      CompletableFuture.supplyAsync(() => {
        val firstFuture: CompletableFuture[ByteBuffer] = first.go(bsonByteBufferEncoding)
        second.go(firstFuture.join()).join()
      })
    future
  }

  override def fuse(boson: Boson) = new BosonFuse(this,boson)
}

File: boson-java/src/main/java/io/zink/boson/impl/BosonFuse.java
package io.zink.boson.impl;

import io.zink.boson.Boson;

import java.nio.ByteBuffer;
import java.util.concurrent.CompletableFuture;

public class BosonFuse implements Boson {
    private Boson first;
    private Boson second;

    BosonFuse(Boson first, Boson second) {
        this.first = first;
        this.second = second;
    }

    @Override
    public CompletableFuture<byte[]> go(byte[] bsonByteEncoding) {
        return CompletableFuture.supplyAsync(() -> {
           /*CompletableFuture<byte[]> firstFuture = first.go(bsonByteEncoding);
           return second.go(firstFuture.join()).join();*/
            return bsonByteEncoding;
        });
    }

    @Override
    public CompletableFuture<String> go(String bsonByteBufferEncoding) {
        return CompletableFuture.supplyAsync(() -> {
            /*CompletableFuture<ByteBuffer> firstFuture = first.go(bsonByteBufferEncoding);
            return second.go(firstFuture.join()).join();*/
            return bsonByteBufferEncoding;
        });
    }

//    @Override
//    public Boson fuse(Boson boson) {
//        return new BosonFuse(this,boson);
//    }
}
